### PR TITLE
[ADD] mail_broker, mail_broker_telegram and mail_broker_whatsapp

### PR DESCRIPTION
**Pull request by olgamarcocb**
The module mail_broker allows to respond chats as a bot, mail_broker_telegram is applied to Telegram and mail_broker_whatsapp to Whatsapp.

TODO:
- [x] Use components
- [ ] Add some tests
- [ ] Merge all the commits in only three ADDs (simplify history)

### CODE REVIEW

File: mail_broker/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+This module allows to respond chats as a bot.
+
+This way, a group of users can respond customers or any other set
+of partners in an integrated way.
+
+It is not intended to be integrated on default chatter as users don't need
```
Comment (JordiBForgeFlow) [inline mail_broker/readme/DESCRIPTION.rst:6]: @olgamarcocb Can you clarify the readme? It's hard to understand

**Comment by rafaelbn** [comment]
Hello @JordiBForgeFlow @etobella is this a WIP? Are you going forward with this one? Thanks 😄

**Comment by eilimrib** [comment]
Neither the Rest API nor the broker seem to be working. Something to do with postMessage in the broker's .js.
image
image

**Comment by etobella** [comment]
Thanks for the Swagger, there was an error there.

For the other error, you cannot post a  message without defining a broker and selecting a chat.

**Comment by etobella** [comment]
Added some improvements in order to be sure that it was sent by whatsapp / telegram / any worker....

**Comment by etobella** [comment]
Passed all the logic to components (we already had them as we needed `base_rest`).

Also, I added some extra security using the secrets in order to ensure that all is fine.

**Comment by marcelsavegnago** [comment]
image

 
 
  File "/mnt/data/odoo-addons-dir/mail_broker_whatsapp/wizards/whatsapp_composer.py", line 56, in action_send_whatsapp
    self._action_send_whatsapp()
  File "/mnt/data/odoo-addons-dir/mail_broker_whatsapp/wizards/whatsapp_composer.py", line 34, in _action_send_whatsapp
    channel = record._whatsapp_get_channel(self.number_field_name, self.broker_id)
  File "/mnt/data/odoo-addons-dir/mail_broker_whatsapp/models/mail_thread.py", line 32, in _whatsapp_get_channel
    channel = broker._get_channel_id(sanitized_number)
AttributeError: 'mail.broker' object has no attribute '_get_channel_id'

**Comment by marcelsavegnago** [comment]
image

**Comment by marcelsavegnago** [comment]
@etobella I really liked the module. Do you plan to migrate it to 14? I tried to do this but I'm having a lot of difficulty because of the MAIL module's javascript refactor.

During my tests on version 13 apparently there is a problem in the get_channel_id method of the mail_broker_whatsapp module which returns that it was not found

broker._get_channel_id(sanitized_number)

https://github.com/tegin/social/blob/efe87717ad9a6482a67fe53c9e6629b4536cfd9c/mail_broker_whatsapp/models/mail_thread.py#L32

**Comment by etobella** [comment]
On the refactoring I moved a function sorry. It should be fixed now :smile: !!

About the migration, I intend to do it in november - december. The hardest art will be JS, as allways :cry:

**Comment by marcelsavegnago** [comment]
@etobella @olgamarcocb.. 

When sending message from whatsapp_composer wizard:


```
  File "/odoo/external-src/oca/social/mail_broker_whatsapp/models/mail_thread.py", line 32, in _whatsapp_get_channel
    channel = broker._get_channel_id(sanitized_number)
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/odoo/src/odoo/http.py", line 639, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/src/odoo/http.py", line 315, in _handle_exception
    raise exception.with_traceback(None) from new_cause
AttributeError: 'mail.broker' object has no attribute '_get_channel_id'
```

https://github.com/tegin/social/blob/15d1bfb13fca51b02016e95deef02f3e117223bc/mail_broker_whatsapp/models/mail_thread.py#L32

**Comment by etobella** [comment]
Thats why i need to add some tests...

**Comment by etobella** [comment]
This is weird... The function is there :thinking: 
https://github.com/OCA/social/pull/817/files#diff-8aed335c0c8c82f838f2001e37dd6ebe45d3015fda98a9e860128619f0b7560bR43

I will try to add some tests soon, this way it could be easier to detect this issues :cry:

**Comment by marcelsavegnago** [comment]
> This is weird... The function is there 🤔

I think I may have taken the wrong branch. I will check and test again. I apologize for the mistake.

**Comment by etobella** [comment]
@marcelsavegnago I added some tests on the function. I will improve it more if I am able :smile:

**Comment by marcelsavegnago** [comment]
@etobella See if it helps.. I put a PR on your fork. https://github.com/tegin/social/pull/1

File: mail_broker/models/mail_broker.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+
+class MailBroker(models.Model):
+    _name = "mail.broker"
+    _inherit = ["collection.base"]
+    _description = "Mail Broker"
+
+    name = fields.Char(required=True)
+    token = fields.Char(required=True)
+    broker_type = fields.Selection([], required=True)
```
Comment (marcelsavegnago) [inline mail_broker/models/mail_broker.py:13]:
```suggestion
    broker_type = fields.Selection([])
```

**Comment by CarlosRoca13** [comment]
I'm trying to configure the telegram broker on my local machine using `ngrok` for the communications with telegram. But when I press on Integrate Webhook, I'm getting 404 

IDK if any extra configuration is needed... I will apreciate a bit of help :sweat_smile: 

image
image

File: mail_broker_telegram/readme/CONFIGURE.rst [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+   https://github.com/tegin/telegram-broker with the TOKEN provided
+
+
+Configure Odoo
+~~~~~~~~~~~~~~
+
+1. Access on debug mode
+2. Access `Settings > Technical Settings > Email > Mail Broker`.
+3. Create a bot and assign the token. Mark it as `Show on App`
+4. Press on `Generate webhook` in order to Open the webhook
```
Comment (CarlosRoca13) [inline mail_broker_telegram/readme/CONFIGURE.rst]:
```suggestion
4. Set a Webhook key and a Webhook secret (use what you want, It's just to allow the comunication with Telegram bot).
5. Press on `Generate webhook` in order to Open the webhook
```

File: mail_broker_telegram/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Limitations
+~~~~~~~~~~~
+
+The Webhook functionality can only be used if your system is accessible from website.
```
Comment (CarlosRoca13) [inline mail_broker_telegram/readme/CONFIGURE.rst]:
```suggestion
The Webhook functionality can only be used if your system is accessible from website and it has `https` certificate.
```

**Review by CarlosRoca13 — APPROVED** [review]
Just documentation comments. Thanks!

**Review by marcelsavegnago — APPROVED** [review]

File: mail_broker_telegram/static/src/xml/thread.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+    <t t-extend="mail.widget.Thread.Message">
+        <t t-jquery=".o_thread_tooltip_broker_container" t-operation="append">
+            <i
+                t-att-class="'o_thread_message_broker_' + message.customer_status + ' fa fa-whatsapp fa-lg'"
+                t-if="message.broker_type == 'whatsapp'"
```
Comment (em230418) [inline mail_broker_telegram/static/src/xml/thread.xml:7]: mail_broker_telegram, but whatsapp is used here

File: mail_broker_whatsapp/services/mail_broker_service.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+from odoo.addons.base_rest import restapi
+from odoo.addons.component.core import Component
+from odoo.addons.mail_broker.services.mail_broker_service import BrokerMethodParams
+
+_logger = logging.getLogger(__name__)
+
+
+class MailBrokerWhatsappService(Component):
+    _inherit = "mail.broker.base.service"
+    _name = "mail.broker.whastapp.service"
```
Comment (em230418) [inline mail_broker_whatsapp/services/mail_broker_service.py:30]: typo
"whastapp"

**Pull request closed without merging**